### PR TITLE
[C10D] Fix coalescedCollective op Flight Recording

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120430

Also noticed and filed https://github.com/pytorch/pytorch/issues/120516 during this work. May land this as is and then test/fix the other varieties of coalesced collectives later.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225 @chauhang

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -2607,11 +2605,10 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collectiveCoalesced(
...
 
-  if (!coalescing_state_ && capture_status == c10::cuda::CaptureStatus::None) {
-    workEnqueue(work);
-  } else {
-    at::cuda::CUDAGraph::dec_pending_event_queries();
-  }
+  workEnqueue(work);
+  // TODO(whc) did I delete a relevant codepath here?
+  // at::cuda::CUDAGraph::dec_pending_event_queries();
+  // }
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2611]: Yes, I think you would need to put back the `capture_status == c10::cuda::CaptureStatus::None` check and the `else` path.

Would be also nice to pack everything:
```
inc_pending_event_queries()
check capture_status
dec_pending_event_queries()
```
inside `workEnqueue` so that we don't have these checks everywhere. (can be in a different PR)

Comment (wconstab) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2622]: I think my reasoning here was that coalescing_state_ should NEVER be active inside this function.  Becuase this function is part of the 'coalescing manager' implementation, which would call its own startCoalescing function later.

If that is correct, i'll add an assertion here to clarify it.

Comment (wconstab) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2622]: i'm not sure how the cudagraph capture path works.  Is it exclusive with coalescing or is it additive?

should the new logic be like this?  I don't understand why the cuda event increment is unconditional and the decrement is conditional.

```
  // Notify graphs before we check the capture status preemptively
  at::cuda::CUDAGraph::inc_pending_event_queries();

  workEnqueue(work);
  if (capture_status != c10::cuda::CaptureStatus::None) {
    at::cuda::CUDAGraph::dec_pending_event_queries();
  }
```

Comment (kwen2501) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2633]: Sorry for the confusion. 

What I meant is like this:
```
  at::cuda::CUDAGraph::inc_pending_event_queries();

  if (capture_status == c10::cuda::CaptureStatus::None) {
    workEnqueue(work);
  } else {
    at::cuda::CUDAGraph::dec_pending_event_queries();
  }
```

It preemptively increasing the pending work number, if this work is actually enqueued, then this is fine.
If this work ends up not being enqueued because CUDA Graph capture is active, then it winds back the pending work number.

Comment (kwen2501) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2633]: You can check elsewhere in the code how `workEnqueue(work)` is conditioned. I guess they are more or less done the same way. Hence I propose putting the above CUDA Graph logic into the `workEnqueue` function itself, so that we don't have the if-else at every call site.

Comment (wconstab) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2633]: hmm.

is it necessary to preemptively increment and then decrement?  or can we just increment when we workEnqueue nad never decrement?

Comment (kwen2501) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2633]: There was a reason. See here:
https://github.com/pytorch/pytorch/pull/110665#issuecomment-1751403208
Cc @eqy

**Review by kwen2501 — APPROVED** [review]
Thanks for the last fix about workEnqueue. LGTM.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -2622,14 +2622,43 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collectiveCoalesced(
...
+  is 0).
+
+  If we are not able to enqueue the work due to capture-in-progress, we finally
+  decrement the counter.
+
+  For this reason we cannot easily move the increment inside workEnqueue unless
+  we also change the semantic of workEnqueue to 'maybeWorkEnqueue'.
+
+  TODO:
+   - Are these assumptions correct? (just pieced this together following
```
Comment (eqy) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2646]: Yes, I think this explanation is accurate; one nit would maybe be editing "(a) capturing is already in progress, so we can not enqueue" to "(a) capturing is already in progress (we cannot enqueue in this case)."
